### PR TITLE
fix: pr_checks now works on macOS

### LINKED ISSUE #41: Mac OS support
I don't have a Mac OS setup. Therefore unable to build dartjni for macos on desktop.

It's build system is different (podspec), and it may take more effort than porting same CMake config to windows.

I searched through github and sourcegraph, and unfortunately not able to find a single podspec that makes use of JNI libraries.

So this would involve finding JNI libraries and listing them in podspec then fixing os-specific errors / warnings..

### PR DESCRIPTION
**Pull request by HosseinYousefi**
Addressing parts of #41.

### CODE REVIEW

**Review by dcharkes — CHANGES_REQUESTED** [review]
Please add MacOS to .github/workflows/test-package.yml

See https://github.com/dart-lang/ffigen/blob/master/.github/workflows/test-package.yml for inspiration on how to setup Mac bots.

File: jnigen/tool/pr_checks.dart [hunk 1]
```diff
@@ -16,26 +16,47 @@
...
-void main() async {
+// Flags
+const _clone = 'clone';
+
+void main(List<String> arguments) async {
+  final parser = ArgParser()
+    ..addFlag(
+      _clone,
+      help:
+          'Git clone will be run and the checks will happen in a temporary directory.',
```
Comment (mahesh-hegde) [inline jnigen/tool/pr_checks.dart:31]: Nit: I think it's conventional to write an imperative statement here.

`Run checks in a cloned copy of the project.`

File: jnigen/tool/pr_checks.dart [hunk 2]
```diff
@@ -60,19 +81,18 @@ void main() async {
...
-        ..chainCommand("dart", ["run", "ffigen", "--config", "ffigen.yaml"])
-        ..chainCallback("compare bindings", () async {
-          final newBindingsText = await ffigenBindings.readAsString();
-          if (newBindingsText != oldBindingsText) {
-            await ffigenBindings.writeAsString(oldBindingsText);
-            throw "new JNI.h bindings differ from old bindings";
-          }
-        });
+  final ffigenCompare = Runner("Generate & Compare FFIGEN bindings", jniPath)
+    ..chainCommand("dart", ["run", "ffigen", "--config", "ffigen.yaml"])
```
Comment (mahesh-hegde) [inline jnigen/tool/pr_checks.dart:85]: I think a small issue is that if the generated bindings are different, old ones are not restored.

If this is problematic, I think adding a `finally` block which writes back the old bindings may solve the issue.

**Review by mahesh-hegde — APPROVED** [review]

**Comment by mahesh-hegde** [comment]
First of all, awesome work @HosseinYousefi.

If I understand correctly jni:setup works on dart standalone.

But since flutter FFIPlugin mechanism is different, we might have to do more things in order to get a flutter desktop using JNI running on macos.

However, it's not that important, because main use on desktop will always be during development use.

Re: checks, you may want to

1. Add a check that runs jni and jnigen tests on macos.

or

refactor existing checks for `jni` and `jnigen` into matrix form. This will require some conditionals because of windows / macos specific things, but make the YAML a little better.

File: jnigen/tool/pr_checks.dart (hunk 1 above)
Comment (HosseinYousefi) [reply jnigen/tool/pr_checks.dart:31]: Much clearer than my statement! Will change it.

File: jnigen/tool/pr_checks.dart (hunk 2 above)
Comment (HosseinYousefi) [reply jnigen/tool/pr_checks.dart:85]: I'll add it for completeness.

File: jnigen/tool/pr_checks.dart [hunk 3]
```diff
@@ -60,19 +81,18 @@ void main() async {
...
+    ..chainCallback("compare bindings", () async {
+      final newBindingsText = await ffigenBindings.readAsString();
+      if (newBindingsText != oldBindingsText) {
+        await ffigenBindings.writeAsString(oldBindingsText);
```
Comment (HosseinYousefi) [inline jnigen/tool/pr_checks.dart:89]: @mahesh-hegde  You're actually already doing it here!

**Comment by dcharkes** [comment]
> Closes https://github.com/dart-lang/jnigen/issues/41.

To call MacOS supported we need a list of things:

* MacOS CLI apps using some jnigen'ed code working
* MacOS Flutter desktop app using some jnigen'ed code working
* MacOS as host for running jnigen on some Java sources
* MacOS as host for Android Flutter app working

Currently, this PR only checks the `package:jni` tests on CI. Which is fine for this PR, but we should keep #41 open until we support it as much as Linux/Windows.

File: jnigen/tool/pr_checks.dart [hunk 4]
```diff
@@ -16,26 +16,46 @@
 import 'dart:async';
```
Comment (dcharkes) [inline jnigen/tool/pr_checks.dart:16]: Should we run this script on the CI to check that we don't break the script accidentally?

Or does this try do things not supported on github bots?

**Review by dcharkes — APPROVED** [review]

**Comment by dcharkes** [comment]
Thanks @HosseinYousefi ! 🎸

**Pull request merged**